crates/ra_ide_api/src/hover.rs

```diff
@@ -118,9 +177,50 @@ impl NavigationTarget {
         // TODO: After type inference is done, add type information to improve the output
         let node = self.node(db)?;
 
+        // FIXME: This is copied from `structure.rs` and should probably
+        // be moved somewhere common
+        fn collapse_ws(node: &SyntaxNode, output: &mut String) {
+            let mut can_insert_ws = false;
+            for line in node.text().chunks().flat_map(|chunk| chunk.lines()) {
+                let line = line.trim();
+                if line.is_empty() {
+                    if can_insert_ws {
+                        output.push_str(" ");
+                        can_insert_ws = false;
+                    }
+                } else {
+                    output.push_str(line);
+                    can_insert_ws = true;
+                }
+            }
+        }
+
+        // FIXME: This is also partially copied from `structure.rs`
+        fn visit_fn(node: &ast::FnDef) -> Option<String> {
+            let mut detail =
+                node.visibility().map(|v| format!("{} ", v.syntax().text())).unwrap_or_default();
+
+            detail.push_str("fn ");
+
+            node.name()?.syntax().text().push_to(&mut detail);
+
+            if let Some(type_param_list) = node.type_param_list() {
+                collapse_ws(type_param_list.syntax(), &mut detail);
+            }
+            if let Some(param_list) = node.param_list() {
+                collapse_ws(param_list.syntax(), &mut detail);
+            }
+            if let Some(ret_type) = node.ret_type() {
+                detail.push_str(" ");
+                collapse_ws(ret_type.syntax(), &mut detail);
+            }
+
+            Some(detail)
+        }
```

Comment:
I think a better approach would be to use `CallInfo` or do something like we do in [CallInfo::new()](https://github.com/rust-analyzer/rust-analyzer/blob/9d22704064e60b0670ee0274b54be968b86f92d6/crates/ra_ide_api/src/call_info.rs#L100). `completion::function_label` is sort of a `Display` for function signatures: we use it for completion and function signature help.

Reply:
`completion::function_label` seems like a good solution for now